### PR TITLE
Container compatibility

### PR DESCRIPTION
**Pull request by marktriggs**
Compatibility changes with the existing ArchivesSpace container model

### CODE REVIEW

File: backend/lib/aspace_json_to_yale_container_mapper.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    result = (try_matching_barcode(container) ||
+              try_matching_indicator_within_series(container) ||
+              try_matching_indicator_within_resource(container))
+
+    if result
+      result
+    else
+      p "UNHANDLED MAPPING - generating new container"
+      p ({:instance => instance, :record => @json})
+      TopContainer.create_from_json(JSONModel(:top_container).from_hash('barcode' => SecureRandom.hex, 'indicator' => (rand() * 10000).to_i.to_s))
```
Comment (marktriggs) [inline backend/lib/aspace_json_to_yale_container_mapper.rb]: Need to think about this... should probably be the default indicator (or the incoming indicator)

File: backend/lib/aspace_json_to_yale_container_mapper.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    end
+  end
+
+
+  def try_matching_barcode(container)
+    # If we have a barcode, attempt to locate an existing top container but create one if needed
+    barcode = container['barcode_1']
+
+    if barcode
+      if (top_container = TopContainer[:barcode => barcode])
```
Comment (marktriggs) [inline backend/lib/aspace_json_to_yale_container_mapper.rb]: Need to scope by repo

File: backend/lib/subcontainer_to_aspace_json_mapper.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  def to_hash
+    result = Hash[JSONModel(:container).schema['properties'].map {|property, _| [property, self.send(property.intern)]}]
+    result['container_locations'] ||= []
+
+    result
+  end
+
+
+  def type_1
+    if container_profile && container_profile.name =~ /Flat Grey/
```
Comment (marktriggs) [inline backend/lib/subcontainer_to_aspace_json_mapper.rb]: Probably bogus.  Should be 'Box'

File: backend/lib/subcontainer_to_aspace_json_mapper.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  include JSONModel
+
+  def initialize(instance_json, instance_obj)
+    @instance_json = instance_json
+    @instance_obj = instance_obj
+  end
+
+
+  def to_hash
+    result = Hash[JSONModel(:container).schema['properties'].map {|property, _| [property, self.send(property.intern)]}]
```
Comment (marktriggs) [inline backend/lib/subcontainer_to_aspace_json_mapper.rb]: Does this map locations?

File: backend/model/mixins/map_to_aspace_container.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    if AppConfig.has_key?(:map_to_yale_container_class)
+      @mapper_to_yale_container ||= Kernel.const_get(AppConfig[:map_to_yale_container_class].intern)
+    else
+      @mapper_to_yale_container ||= AspaceJsonToYaleContainerMapper
+    end
+  end
+
+
+
+  def self.mapper_to_aspace_json=(clz)
```
Comment (marktriggs) [inline backend/model/mixins/map_to_aspace_container.rb]: Is this needed?  Looks wrong.

File: backend/spec/model_compatibility_spec.rb [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+                                                                                           'ref' => location_2.uri)
+                                                ]
+                                              })
+
+    expect {
+      accession = create_accession({"instances" => [instance]})
+    }.to raise_error(ValidationException)
+  end
+
+
```
Comment (marktriggs) [inline backend/spec/model_compatibility_spec.rb]: Test location mapping here

**Comment by payten** [comment]
:rocket:

**Pull request merged**